### DISCUSSION TITLE
Ability to reuse `type alias` in another assembly with apply all encapsulation limitation

### DISCUSSION THREAD

**Discussion by soroshsabz** [dotnet/csharplang#7253] [Language Ideas]
ITNOA

As you know in C# language we cannot use [`using alias`](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/using-directive#using-alias) in another assembly or in another `using alias`.

In another word, we cannot using `using alias` as normal type, that transparent type declaration for users. this feature exist in C++ for many years, and it is very useful specially in generic complex type aliasing.

for example I want to have something like below

```
namespace UsingTest
{

    public interface IUsingTest
    {
        using MyType = System.Collections.Generic.List<System.Collections.Generic.Dictionary<string, int>>;
        MyType GetList();
    }
}
```

and in another `class` (**in same assembly or in another assembly**) we want to use this type is natural way like below

```
namespace ConsoleTest
{
    internal class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("Hello, World!");

            UsingTest.IUsingTest.MyType foo = UsingImplementaiton.GetList();
        }
    }
}
```

or in another use case, I want to add scope to `using alias` in class inheritance model for example

```
class Foo
{
   public using MyType = List<int>;
   protected using OnlyChildType = Dicitonary<int, MyType>;
};
```

the `public using alias`  provide this type globally for all users of `Foo` class, but `protected using alias` provide this type only for derived classes.

in the end, I think if `using alias` behave like `using` in C++ we can provide a better abstraction and encapsulation and decoupling of codes.

thanks

**Comment by HaloFour** [comment]
I believe that this will fall out with extensions: #5497

Explicit extensions specifically allow for a zero-cost abstraction for a subtype of any existing type that can provide type safety and additional functionality through extension members.

**Reply by soroshsabz** [reply]
Why implement this with extensions? I think it is natural expected behavior of type aliasing.

**Reply by HaloFour** [reply]
Because extensions are already being designed in the language and carry the metadata required in order to expose this between assemblies.  As aliases are nothing more than syntax candy you would need a completely different mechanism to define them across assembly boundaries.

Extensions are also substantially more capable.  They will allow actual type safety, unlike aliases where the 5000 different aliases for `int` are all interchangeable (and IMO a very good reason to _not_ implement them like C++).  They will also allow extension members to extend the functionality of that type for domain purposes.

**Reply by soroshsabz** [reply]
As you can say, extensions is for extending functionality, but as you see in this requirement, we just want to renaming type for better describing the usage, and prevent changing chain, if I want to change the type in future. So I think this feature is fundamentally different than extension feature.

and for your example in _unlike aliases where the 5000 different aliases for int are all interchangeable_ , , I think the key note is type aliasing spirit and nature. and the core requirements for type aliasing is very like variable (type instantiation) that we want to make a name for specific purpose and play with it. so 5000 different aliases Not only is it not a bad thing, it is also a very good thing.

sorry for bad English (it is not my native language) :(

**Reply by HaloFour** [reply]
> So I think this feature is fundamentally different than extension feature.

Maybe, but if the same functionality falls out with extensions there's no reason to design it twice.

> I think the key note is type aliasing spirit and nature. and the core requirements for type aliasing is very like variable (type instantiation) that we want to make a name for specific purpose and play with it. 

I disagree.  I think the name is pointless if there is no type safety to go with it.  However, extensions will allow for this as well.

**Reply by soroshsabz** [reply]
@CyrusNajmabadi what is your opinion?

**Reply by kasajian** [reply]
@HaloFour  I very much agree with the statement that without type-safety, it's not as useful.   In C typdefs have this problem, and we have to play games with hidden structs (such as how HWND is defined in win32 strict mode).   I think there are types when you want simply have a different name for the existing type.   In C this was very useful for instance I want to have type INT32 and INT16 to always be 32-bit and 16-bit, and typedef would map to the appropriate type for the particular hardware.

I can't think of very good examples where that's needed in C#.   Some of this reminds me Nominal types in Typescript where even two independent interfaces are basically the same type as far as the compiler is concerned, if their struct is identical.

What you're talking about our ValueObjects, and probably the #1 feature missing for me in C#.   What you really want is something like Vogen but built-in: https://github.com/SteveDunn/Vogen

**Reply by Richiban** [reply]
> Explicit extensions specifically allow for a zero-cost abstraction for a subtype of any existing type that can provide type safety 

Do extensions allow "constructors"? I.e. for a `public explicit extension SpecialPerson for Person` can I make the conversion from `Person` to `SpecialPerson` fail with an exception if certain conditions are not met?

**Reply by HaloFour** [reply]
See: #8161

Per the Discord conversations, the team is very interested in "validation" for explicit extensions.

**Reply by Richiban** [reply]
> See: #8161
> 
> Per the Discord conversations, the team is very interested in "validation" for explicit extensions.

That's great to hear, cheers!

**Comment by soroshsabz** [comment]
related to #4284 
related to #1239
